Engine/Platform/System/Posix/CYBPosixFile.cpp

```diff
@@ -0,0 +1,160 @@
+//! @file CYBPosixFile.cpp Implements CYB::Platform::System::File for Posix
+#include "CYB.hpp"
+
+using namespace CYB::Platform::Posix;
+
+void CYB::Platform::System::Implementation::File::Init(const System::Path& APath, const API::File::Mode AMode, API::File::Method& AMethod) {
+
+	if (AMode == API::File::Mode::READ && AMethod == API::File::Method::TRUNCATE)
+		throw Exception::Violation(Exception::Violation::INVALID_PARAMETERS);
+
+	auto Flags([&]() {
+		switch (AMode) {
+		case API::File::Mode::READ:
+			return O_RDONLY;
+		case API::File::Mode::WRITE:
+			return O_WRONLY;
+		case API::File::Mode::READ_WRITE:
+			return O_RDWR;
+		default:
+			throw Exception::Violation(Exception::Violation::INVALID_ENUM);
+		}
+	}());
+
+	switch (AMethod) {
+	case API::File::Method::ANY:
+	case API::File::Method::CREATE:
+		Flags |= O_CREAT | O_EXCL;
+		break;
+	case API::File::Method::EXIST:
+		break;
+	case API::File::Method::TRUNCATE:
+		Flags |= O_CREAT | O_TRUNC;
+		break;
+	default:
+		throw Exception::Violation(Exception::Violation::INVALID_ENUM);
+	}
+
+	FDescriptor = Core().FModuleManager.FC.Call<Modules::LibC::open>(APath().CString(), Flags, S_IRUSR | S_IWUSR);
+
+	if (FDescriptor == -1) {
+		const auto Error(errno);
+		switch (Error) {
+		case EEXIST:
+			if (AMethod == API::File::Method::ANY) {
+				AMethod = API::File::Method::EXIST;
+				Init(APath, AMode, AMethod);
+				return;
+			}
+		case EISDIR:
+			throw Exception::SystemData(Exception::SystemData::FILE_EXISTS);
+		case ENOTDIR:
+		case ENOENT:
+			throw Exception::SystemData(Exception::SystemData::FILE_NOT_FOUND);
+		case EACCES:
+		default:
+			if (AMode == API::File::Mode::READ)
+				throw Exception::SystemData(Exception::SystemData::FILE_NOT_READABLE);
+			else
+				throw Exception::SystemData(Exception::SystemData::FILE_NOT_WRITABLE);
+		}
+	}
+
+	//Check it's actually a file because otherwise we aren't supposed to have it open
+	if (!(S_ISREG(StatFD().st_mode))) {
+		Close();
+		throw Exception::SystemData(Exception::SystemData::FILE_EXISTS);
+	}
+
+	if (AMethod == API::File::Method::ANY)
+		AMethod = API::File::Method::CREATE;
+}
+ 
+CYB::Platform::System::File::File(Path&& APath, const Mode AMode, const Method AMethod) :
+	FPath(std::move(APath)),
+	FOpenMode(AMode),
+	FOpenMethod(AMethod)
+{
+	Init(FPath, AMode, FOpenMethod);
+}
+
+CYB::Platform::System::Implementation::File::File(File&& AMove) noexcept :
+	FDescriptor(AMove.FDescriptor) {
+	AMove.FDescriptor = -1;
+}
+
+CYB::Platform::System::File& CYB::Platform::System::File::operator=(File&& AMove) noexcept {
+	Close();
+	FDescriptor = AMove.FDescriptor;
+	AMove.FDescriptor = -1;
+	FOpenMode = AMove.FOpenMode;
+	FOpenMethod = AMove.FOpenMethod;
+	return *this;
+}
+
+void CYB::Platform::System::Implementation::File::Close(void) const noexcept {
+	if (FDescriptor != -1)
+		Core().FModuleManager.FC.Call<Modules::LibC::close>(FDescriptor);
+}
```

Comment:
Flush support has been intentionally left out as it represents a lower level we are trying to abstract away. Assume data is written once the calls are completed.